site/_posts/2023-02-03-single-sign-on-laravel-fusionauth.md

```diff
@@ -0,0 +1,345 @@
+
+Now, configure your application with the necessary settings to interact with the FusionAuth service, by adding a new entry to the array located at `config/services.php`:
+
+```php
+'fusionauth' => [
+    'client_id' => env('FUSIONAUTH_CLIENT_ID'),
+    'client_secret' => env('FUSIONAUTH_CLIENT_SECRET'),
+    'redirect' => env('FUSIONAUTH_REDIRECT_URI'),
+    'base_url' => env('FUSIONAUTH_BASE_URL'),
+    'tenant_id' => env('FUSIONAUTH_TENANT_ID'),
+],
+```
+
+### Configuring environment variables
+
+Instead of putting the values directly there, you should use environment variables, which is a good practice from both maintenance and security standpoints, as described in the [12 Factor App](https://12factor.net/config) methodology and in several others. Being so, alter your `.env` file to include those entries:
+
+```ini
+# Paste both client Id and secret for your application
+FUSIONAUTH_CLIENT_ID=<APP CLIENT ID FROM FUSIONAUTH>
+FUSIONAUTH_CLIENT_SECRET=<APP CLIENT SECRET FROM FUSIONAUTH>
+
+# Specify a tenant Id or leave this blank for the default tenant
+FUSIONAUTH_TENANT_ID=<ID FOR YOUR TENANT>
+
+# Replace http://localhost:9011 with the address where the FusionAuth application is running
+FUSIONAUTH_BASE_URL=http://localhost:9011
+
+# Replace http://localhost with the address for your PHP application, if necessary
+# We'll create the route for `/auth/callback` later##
+FUSIONAUTH_REDIRECT_URI=http://localhost/auth/callback
+```
+
+### Creating new fields in the User model
+
+Migrations are the code-based version control for your database. Instead of having to share SQL queries with your coworkers so they can create a table or add a column everytime there is a change in the database, you can write code to programmatically do this when running a command.
+
+In Laravel, to create a new migration you just need to run the `artisan make:migration` and inform a meaningful name for it. In the next line, we'll create a migration called `add_fusionauth_fields_user_table`:
+
+```shell
+$ ./vendor/bin/sail artisan make:migration add_fusionauth_fields_user_table
+```
+
+Then, go to your `database/migrations` folder, where you'll find a file there with the name you created and prefixed by today's date (e.g `2023_02_03_123000_add_fusionauth_fields_user_table.php`).
+
+Edit that file to add three new columns to the `users` table: `fusionauth_id`, `fusionauth_access_token` and `fusionauth_refresh_token` and allow `NULL` values in the `password` column. You'd want to save the access token in your database to make requests to the FusionAuth API on behalf of the user. The refresh token is used to exchange an expired access token with a new one (but you'd still have to implement this yourself, as Socialite still doesn't do it). This way, users don't need to log in again to your FusionAuth instance as long as the access token is still valid.
```

Comment:
```suggestion
Edit that file to add three new columns to the `users` table: `fusionauth_id`, `fusionauth_access_token` and `fusionauth_refresh_token` and allow `NULL` values in the `password` column. You'd want to save the access token in your database to make requests to the FusionAuth API or other APIs on behalf of the user. The refresh token is used to exchange an expired access token with a new one (but you'd still have to implement this yourself, as Socialite still doesn't do it). This way, users don't need to log in again to your FusionAuth instance as long as the refresh token is still valid.
```

Reply:
If they had another laravel API (or other custom API) that accepted FusionAuth access tokens, you could use them to access those APIs.